pkg/pillar/cmd/nim/controllerdns.go

```diff
@@ -105,42 +105,60 @@ func (n *nim) controllerDNSCache(
 		n.Log.Warnf("%s exists but removing failed: %+v", tmpHostFileName, err)
 	}
 
-	var newhosts []byte
+	dnsResponses := n.resolveWithPorts(string(controllerServer))
+
+	lookupIPaddr := n.writeHostsFile(dnsResponses, etchosts, controllerServer)
+	if lookupIPaddr != "" {
+		n.Log.Tracef("append controller IP %s to /etc/hosts", lookupIPaddr)
+	}
+
 	var ttlSec int
-	var lookupIPaddr string
 
-	dnsResponses := n.resolveWithPorts(string(controllerServer))
+	if len(dnsResponses) > 0 {
+		lastDNSResponse := dnsResponses[len(dnsResponses)-1]
+		ipaddrCached = lastDNSResponse.IP.String()
+		ttlSec = getTTL(time.Duration(lastDNSResponse.TTL))
+		return ipaddrCached, ttlSec
+	} else {
+		return "", ttlSec
+	}
+}
+
+func (n *nim) writeHostsFile(
+	dnsResponses []devicenetwork.DNSResponse,
+	etchosts, controllerServer []byte,
+) string {
+	return n.writeHostsFileToDestination(dnsResponses, etchosts, controllerServer, etcHostFileName)
+}
+
+func (n *nim) writeHostsFileToDestination(
+	dnsResponses []devicenetwork.DNSResponse,
+	etchosts, controllerServer []byte,
+	destination string,
+) string {
+	var newhosts []byte
+
+	var lookupIPaddr string
 
 	if len(dnsResponses) == 0 {
 		newhosts = append(newhosts, etchosts...)
 	} else {
-		ttlSec = int(dnsResponses[0].TTL)
 		lookupIPaddr = dnsResponses[0].IP.String()
```

Comment:
Is the old/current code only writing one IP address to the hosts file? Or does it write all? 
You could make this loop over dnsResponses and Sprintf ine line for each if that is the behavior we want.

Reply:
The current master only writes one IP address to the hosts file; this branch writes all IP addresses to the hosts file.

I think this is the behavior we want; I saw that f.e. ping would choose the ip address that is reachable from the hosts file if there is more than one.